yoga v3 blog post (#928)
the-guild-org/website: website/pages/blog/graphql-yoga-v3.mdx

---
title: GraphQL Yoga v3
tags: [graphql, svelte]

> authors: [saihaj, laurin, arda]

Reviewer: denis

---

> title: Announding GraphQL Yoga v3

Reviewer 2: ```suggestion
title: Announcing GraphQL Yoga v3
```

tags: [graphql, nodejs, typescript, yoga]
authors: [saihaj, arda, enisdenjo, laurin]
date: 2022-11-15
description: The new version of GraphQL Yoga is out! Learn what has improved and what is new!
image: /x
thumbnail: /x
---

After months of experimentation and learning from production use cases we are back again with a ground breaking release for GraphQL Yoga.

## The Yoga v3 Experience

The goal with Yoga is to empower developers, improve the developer facing API and make caching, security and real-time easy and run anywhere.
Let's highlight some of the groundbreaking new features and improvements.

> ### One Package to rule them all

Reviewer 2: This is a bit aggressive 👀 

Maybe consider the following?

```suggestion
### One Package, runs everywhere
```

From GraphQL Yoga v2 we figure that using the `fetch` API as standard is the way to go.

> Now GraphQL Yoga has a single NPM package for all environments `graphql-yoga` instead of `@graphql yoga/common` and `@graphql-yoga/node`.

Reviewer 3: ```suggestion
Now GraphQL Yoga has a single NPM package for all environments `graphql-yoga` instead of `@graphql-yoga/common` and `@graphql-yoga/node`.
```

> We [created a library that lets you create cross-platform HTTP servers](https://www.the-guild.dev/blog/fetch-for-servers) which GraphQL Yoga v3 uses under the hood.

Reviewer 4: plz remove `www` from the link

> We [created a library that lets you create cross-platform HTTP servers](https://www.the-guild.dev/blog/fetch-for-servers) which GraphQL Yoga v3 uses under the hood.

Author: ```suggestion
We [created a library that lets you create cross-platform HTTP servers](https://the-guild.dev/blog/fetch-for-servers) which GraphQL Yoga v3 uses under the hood.
```

> We [created a library that lets you create cross-platform HTTP servers](https://www.the-guild.dev/blog/fetch-for-servers) which GraphQL Yoga v3 uses under the hood.

Author: ```suggestion
We [created a library that lets you create cross-platform HTTP servers](https://the-guild.dev/blog/fetch-for-servers) which GraphQL Yoga v3 uses under the hood.
```

> We [created a library that lets you create cross-platform HTTP servers](https://www.the-guild.dev/blog/fetch-for-servers) which GraphQL Yoga v3 uses under the hood.

Author: ```suggestion
We [created a library that lets you create cross-platform HTTP servers](https://the-guild.dev/blog/fetch-for-servers) which GraphQL Yoga v3 uses under the hood.
```

[Learn more](https://the-guild.dev/graphql/yoga-server/docs)

### The Makeover

> GraphQL Yoga v3 comes with the brand new GraphiQL 2.0

Author: ```suggestion
GraphQL Yoga v3 comes with the brand new GraphiQL 2.0.

The development of GraphiQL 2.0 has been a long ongoing process and we are so happy to finally replace the outdated and old-fashioned GraphiQL IDE with the new fancy and snappy version 2.

Thanks to all parties involved that made this happen.
```

Author: @dotansimha like this?

![GraphiQL](https://raw.githubusercontent.com/graphql/graphiql/main/packages/graphiql/resources/graphiql.png)

[Learn more](https://the-guild.dev/graphql/yoga-server/docs/features/graphiql)

### 100% compliant

> As of today, GraphQL Yoga is only GraphQL servers in the ecosystem that fully complies with GraphQL over HTTP specification.

Reviewer 4: `... is the only GraphQL server in the...`

Reviewer 4: Link in `GraphQL over HTTP specification`

Reviewer 5: Is the only **JavaScript** GraphQL server. [HotChocolate is fully compliant too.](https://github.com/graphql/graphql-http/blob/master/implementations/hotchocolate/README.md)

Author: ```suggestion
As of today, GraphQL Yoga is the only GraphQL server in the JavaScript ecosystem that fully complies with GraphQL over HTTP specification.
```

Author: Do you guys know that god invented the suggestion feature?

| Name                     | Pass ✅ | Warning ⚠️ | Errors ❌ |
| ------------------------ | ------- | ---------- | --------- |
| `graphql-yoga`           | 72      | 0          | 0         |
| `apollo-server@3.10.2`   | 42      | 29         | 1         |
| `express-graphql@0.12.0` | 36      | 35         | 1         |
| `graphql-helix@1.13.0`   | 39      | 32         | 1         |

> | `mercurius@11.0.0`       | 41      | 30         | 1         |

Reviewer 5: The table is outdated, see https://github.com/graphql/graphql-http#servers.

[Learn more](https://github.com/graphql/graphql-http/tree/master/implementations/graphql-yoga)

> ### Yoga as a Gateway

Reviewer 2: The paragraph below mainly talks about performance, not Gateway

> GraphQL Yoga has been built-in real-world conditions within our projects (e.g. GraphQL Mesh, GraphQL Hive) and with some of our clients and performance is highly prioritized.

Reviewer 4: something is off with this sentence 😆

> GraphQL Yoga has been built-in real-world conditions within our projects (e.g. GraphQL Mesh, GraphQL Hive) and with some of our clients and performance is highly prioritized.

Reviewer 4: ```suggestion
GraphQL Yoga has been built in real-world conditions within our projects (e.g. GraphQL Mesh, GraphQL Hive) and with some of our clients, and performance is highly prioritized.
```

We continuously keep track and improve performance.
We have a rigorous test suite to avoid any regressions.
The ease of using GraphQL Yoga anywhere and performance make it a great candidate to run as a gateway.
Yoga is compliant with Apollo Federation spec so you can use it for federated services.

[Learn more](https://the-guild.dev/graphql/yoga-server/docs/features/apollo-federation)

> ### Defer and Stream

Reviewer 2: ```suggestion
### Defer and Stream support
```

})

const server = createServer(yoga)

server.listen(4000, () => {
  console.info('Server is running on http://localhost:4000/graphql')
})
```

[Learn more](https://the-guild.dev/graphql/yoga-server/docs/features/defer-stream)

### Not only GraphQL but also REST

There are many reasons why one would want a REST endpoint.
The Sofa API plugin makes it easy to convert any GraphQL API to REST API and in Yoga v3.

```ts
import { createYoga } from 'graphql-yoga'
import { createServer } from 'node:http'
import { useSofaWithSwaggerUI } from '@graphql-yoga/plugin-sofa'
import { schema } from './schema.js'

export const yoga = createYoga({
  schema,
  plugins: [
    useSofaWithSwaggerUI({
      basePath: '/rest',
      swaggerUIEndpoint: '/swagger',
      servers: [
        {
          url: '/', // Specify Server's URL.
          description: 'Development server'
        }
      ],
      info: {
        title: 'Example API',
        version: '1.0.0'
      }
    })
  ]
})

const server = createServer(yoga)
server.listen(4000, () => {
  console.info('Server is running on http://localhost:4000/graphql')
})
```

[Learn more](https://the-guild.dev/graphql/yoga-server/docs/features/sofa-api)

> ### Response Cachning

Reviewer 5: ```suggestion
### Response Caching
```
